util/util.go

```diff
@@ -331,3 +333,27 @@ func NewLoggingMiddleware(writer io.Writer) func(http.Handler) http.Handler {
 func SanitizeLogEntry(entry string) string {
 	return strings.Replace(strings.Replace(entry, "\n", "", -1), "\r", "", -1)
 }
+
+func randomCharacter() string {
+	for i := 0; i < 5; i++ {
+		character, err := GetRandomString(1)
+		if err != nil {
+			continue
+		}
+		return character
+	}
+	return ""
+}
+
+func NewID() string {
+	newID, err := shortid.Generate()
+	if err != nil {
+		newID = uuid.New().String()
+	} else {
+		// remove underscores and hyphens from short ID. The hypens will remain
+		// if we are forced to fall back to uuid4.
+		newID = strings.Replace(newID, "_", randomCharacter(), -1)
+		newID = strings.Replace(newID, "-", randomCharacter(), -1)
```

Comment:
I'm no expert here. But I assume that this might compromise the uniqueness of the ids?! You can define the alphabet within the shortid lib, but if you remove '-' and '_' it what other chars instead. 
https://github.com/teris-io/shortid/blob/master/shortid.go#L85 
The algorithm seems to always want 64 distinct chars. But I thought replacing '_' and '-' by '%' and '$' is not much better ..

I really would want to avoid some non-unique issue here.

Reply:
> I really would want to avoid some non-unique issue here.

indeed. It erred out on OpenStack when the name included an underscore :smile:. The need for 64 unique characters leaves us with few options. We can leave the hyphen and maybe add a ```+``` ? Something that has a lowest chance of failing anywhere garm may be used. Will have to look at the other clouds and what limitations there are in regards to names.

Reply:
I'll remove this hack as it made my eye twitch as well when I wrote it.

Reply:
Funny.. my openstack is fine with '_'

Reply:
ohh wait. My bad. The error was on LXD:

```bash
Jan 27 14:18:20 garm garm[11696]: 2023/01/27 14:18:20 failed to add instance to provider: creating instance: creating instance: creating instance: Invalid instance name: Name can only contain alphanumeric and hyphen characters
```

The same thing seems to be true for Azure:

https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules (underscores are not allowed)

From the Azure docs above, in regards to ```VirtualMachines```:

```
Can't use spaces, control characters, or these characters:
~ ! @ # $ % ^ & * ( ) = + _ [ ] { } \ | ; : . ' " , < > / ?

Windows VMs can't include period or end with hyphen.

Linux VMs can't end with period or hyphen.

Which is depressing. Need to find a way to get around this.
```

Reply:
I think I'll leave the current dictionary alone, and just retry generating a new ID until I get one without those two characters.

Reply:
actually, just the one. The underscore.

Reply:
okay. Added a retry (100x). If we can't get a short ID without an underscore in it after 100 tries, we fall back to UUID. Should be fast enough even with this retry.

Reply:
Uh, sorry for being picky now, but trying until no underscore comes is the expensive version of directly replacing it randomly. From an entropy perspective it has the same effect. Should we go back to uuid or maybe a config toggle uuid/shortid?

Reply:
Or we use a different library?!

Reply:
Was looking now for alternatives now. I like the short ID idea.

Reply:
hmm. I wonder if we could encode the short ID (or even the UUID) into ```base62```. It should solve our problem and shorten even the regular UUID considerably. Uniqueness should be preserved. What do you think?

Reply:
Looks ok. We get a slightly longer ID as a consequence (between 12 and 14), and the UUID becomes 22 characters long.

Reply:
but base 64 contains '/' and '+' and also '=' for padding at the end. 

I also like the shortid though, :smile:

Reply:
base 62 does not :smile: should be alphanumeric (including upper/lower case)

Reply:
https://en.wikipedia.org/wiki/Base62

Reply:
Looks like this:

![Screenshot from 2023-01-29 17-51-27](https://user-images.githubusercontent.com/903160/215338234-852a84ee-e0df-4261-95c6-1a5d59f4ec21.png)

Reply:
okay .. never heard of that ;)
sgtm